doc/dotNET SDK Proposal.md

```diff
@@ -0,0 +1,232 @@
+
+### Workflow Host Configuration
+
+The user-facing foundation of the SDK is the host. It is configured in a manner that closely follows the conventions for configuring an ASP.NET Core application host ([basic example](../src/Samples/Basic/Temporal.Samples.DesignProcess/Part1_1_TypicalWorkflow.cs#L73-L89) | [more examples](../src/Samples/Basic/Temporal.Samples.DesignProcess/Part4_1_BasicWorkflowUsage.cs#L67-L202) | [another example](../src/Samples/Basic/Temporal.Samples.DesignProcess/Part4_2_BasicWorkflowUsage_MultipleWorkers.cs#L108-L158)).
+
+Users provide workflow implementations and activity implementations in one of two ways:
+ - Pattern based: Entry points of workflow and activity implementations follow specific, well known patterns. Developers use attributes to provide metadata about such entry point to the host. The host automatically loads, and executes workflows and activities. This approach allows writing workflows (and activities) with hardly any scaffolding. 
+- Interface based: Developers implement specific interfaces to create workflows (and activities), and point the host to these implementations. The host automatically loads, and executes workflows and activities. This approach requires a little more scaffolding. However, it also allows additional functionality and helps addressing some advanced scenarios (e.g. dynamic workflow APIs).
+
+### Basic Workflow Execution
+
+Consider a sample workflow of type `SomeWorkflow` that runs a loop executing `SomeActivity` until a certain condition is met. When such condition is eventually met, the workflow is notified via a `NotifyConditionMet` signal.
+
+At application startup, the application is configured to host a worker that can execute workflows (and activities). Then, the worker is configured to poll `SomeTaskQueue` for work items, and to support execution of `SomeWorkflow` instances:
+
+```cs
+public static void Main(string[] args)
+{     
+    IHost appHost = Host.CreateDefaultBuilder(args)
+            .UseTemporalWorkerHost()
+            .ConfigureServices(serviceCollection =>
+            {
+                serviceCollection.AddTemporalWorker()
+                        .Configure(temporalWorkerConfig =>
+                        {
+                            temporalWorkerConfig.TaskQueueMoniker = "SomeTaskQueue";
+                        });
+
+                serviceCollection.AddWorkflowWithAttributes<SomeWorkflow>();
+            })
+            .Build();
+
+    appHost.Run();
+}
+```
+
+We use the `WorkflowAttribute` to mark the class that implements the workflow and to point to the method that implements the main workflow logic. The `WorkflowSignalHandler` is used to specify the method that handles `NotifyConditionMet`-signals:
+
+```cs
+[Workflow(runMethod: nameof(MainAsync))]
+public class SomeWorkflow
+{
+    private bool _isConditionMet = false;
+
+    public async Task MainAsync(WorkflowContext workflowCtx)
```

Comment:
Rust has a concept of task-local storage which I could've used for this in the prototype SDK (maybe something similar in .NET) but I think it explicitly being passed in is actually preferable in a lot of ways. Explicit over implicit and all that

Reply:
> I think it explicitly being passed in is actually preferable in a lot of ways. Explicit over implicit and all that

I would agree with this normally, but this is environment-level stuff. You should not have to accept a certain parameter in your workflow to call an activity from a workflow IMO. We don't require that in TS, and we don't require it in Java, we only do it in Go because they explicitly chose that paradigm and goroutine-local is not explicitly supported.

It's like saying telling a developer they have to accept a special "http context" before they are allowed to make HTTP calls or an IO context to write to disk. It may seem neat from a capabilities perspective (and some langs like Haskell require such explicit environment contexts/wrappers), but it's really not simple from an environment perspective.

Also, lack of magic required params representing the environment allow you to reuse the function in other settings such as for client calls in dynamic invocation ala Java proxies (e.g. `var workflowClient = client.newWorkflowClient<MyWorkflow>(); workflowClient.MyMethodWithoutContext("foo");`).

Reply:
There is an "async local" in .NET. However, I prefer the explicit approach. Magical context parameters can make unit testing harder. Sharing an interface for client-side and server side is hard anyway ([example](https://github.com/chronoguy/sdk-dotnet/blob/proposal-01/src/Samples/Basic/Temporal.Samples.DesignProcess/Part1_5_AttributesAndInterfaces.cs)). In case of doubt, we should consider that adding a magical context parameter is always possible in a future version, but it is not so easy to take it away. So we can always start with explicit and then see how much demand we get from the community.

Reply:
I think Java or TypeScript is the thing to emulate here, not Go. I don't think we should have an explicit context. I don't see any reason to deviate from the Java SDK here. I am not sure it is best to require it even at first as it severely affects some of the API such as execute child workflow. I will defer to others here, but from a workflow perspective, I consider this as much a part of the environment as the filesystem and network.

It feels like we are adding more wordiness/explicitness for this environment just because of a general "explicit over implicit" programming paradigm, when really we should be striving for simplicity. The question to ask is: why didn't we require the workflow context in all parameters in TypeScript? Why can I just call `sleep` instead of `context.sleep` or `sleep(context)`? Are y'all saying we made a mistake in the TypeScript SDK there? If we're changing tack, we need to be clear we want to as other SDKs come up.

Reply:
Yeah, that all makes sense. I agree if it can be accessed without an extra param in a way that's safe and possibly overrideable for testing, then we should do that to stay consistent and keep the signatures smaller.

Reply:
Implicit contexts tend to come with a lot of difficulties:
- they are harder to impossible to mock-up in test code
- libraries that use such implicit context cannot make it a part of their API signature
- they make it very hard to evolve the execution environment in a way that changes the synchronization context (e.g. what if in some future version we would want to permit other asynchrony models then "`await`"?
- these "per logical execution thread" singletons suffer from all the same issues as common singletons ([search](https://www.google.com/search?q=problems+with+singleton&oq=problems+with+singleton)).

That is quite high cost to save a parameter. In fact, I would strongly advocate to moving away from that pattern in new SDKs.

However, the most decisive point from my perspective is that a magical context can always be added in a future version if the need becomes overwhelming. But it cannot be removed. So let's not design ourselves into a corner we do not want to be in.

Either way, this is a very important topic and I will dedicate a section to in in the next doc verson.